test/optimizer/plugins/AddPathRulesOptimizer.test.ts

```diff
@@ -0,0 +1,169 @@
+describe('optimizer', () => {
+  describe('add_rule_paths', () => {
+    let myPackage: Package;
+
+    beforeEach(() => {
+      myPackage = new Package();
+    });
+
+    it('should add path rules based on repeated parent paths', () => {
+      // Instance: ChainsawPatient
+      // InstanceOf: Patient
+      // Usage: #example
+      // * address.city = "Tokyo"
+      // * address.country = "JP"
+      const instance = new ExportableInstance('ChainsawPatient');
+      instance.instanceOf = 'Patient';
+      instance.usage = 'Example';
+      const tokyoCity = new ExportableAssignmentRule('address.city');
+      tokyoCity.value = 'Tokyo';
+      const japanCountry = new ExportableAssignmentRule('address.country');
+      japanCountry.value = 'JP';
+      instance.rules.push(tokyoCity, japanCountry);
+
+      // Instance: ChainsawPatient
+      // InstanceOf: Patient
+      // Usage: #example
+      // * address
+      // * address.city = "Tokyo"
+      // * address.country = "JP"
+      const addressPathRule = new ExportablePathRule('address');
+      const expectedPathRule = cloneDeep(addressPathRule);
+      const expectedTokyoCity = new ExportableAssignmentRule('address.city');
+      expectedTokyoCity.value = 'Tokyo';
+      const expectedJapanCountry = new ExportableAssignmentRule('address.country');
+      expectedJapanCountry.value = 'JP';
+
+      myPackage.add(instance);
+      optimizer.optimize(myPackage);
+      expect(instance.rules).toEqual([expectedPathRule, expectedTokyoCity, expectedJapanCountry]);
+    });
+
+    it('should add path rules if the parent paths are numbered', () => {
+      // Instance: ChainsawPatient
+      // InstanceOf: Patient
+      // Usage: #example
+      // * address[0].city = "Shanghai"
+      // * address[0].country = "CN"
+      // * address[1].city = "Sydney"
+      // * address[1].country = "AU"
+
+      const instance = new ExportableInstance('ChainsawPatient');
+      instance.instanceOf = 'Patient';
+      instance.usage = 'Example';
+      const tokyoCity = new ExportableAssignmentRule('address[0].city');
+      tokyoCity.value = 'Tokyo';
+      const japanCountry = new ExportableAssignmentRule('address[0].country');
+      japanCountry.value = 'JP';
+      const sydneyCity = new ExportableAssignmentRule('address[1].city');
+      sydneyCity.value = 'Sydney';
+      const australiaCountry = new ExportableAssignmentRule('address[1].country');
+      australiaCountry.value = 'AU';
+      instance.rules.push(tokyoCity, japanCountry, sydneyCity, australiaCountry);
+
+      // Instance: ChainsawPatient
+      // InstanceOf: Patient
+      // Usage: #example
+      // * address[0]
+      // * address[0].city = "Shanghai"
+      // * address[0].country = "CN"
+      // * address[1]
+      // * address[1].city = "Syndney"
+      // * address[1].country = "AU"
+
+      const expectedPathRule0 = new ExportablePathRule('address[0]');
+      const expectedTokyoCity = cloneDeep(tokyoCity);
```

Comment:
Similar to above, for this test and the ones below it, can we just use the rules directly and not clone them? For rules that don't change, it should still be okay to just use the rule directly.

Reply:
Agreed. In some tests, the clone is helpful when we change the data in the expected rules -- but it doesn't seem we're doing that here.

Reply:
Tidied up the tests in 5df2936.